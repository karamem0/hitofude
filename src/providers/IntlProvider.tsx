//
// Copyright (c) 2023-2025 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/hitofude/blob/main/LICENSE
//

import React from 'react';

import { IntlProvider as Provider, createIntl } from 'react-intl';
import en from '../translations/en.json';
import ja from '../translations/ja.json';

const translations: { [key: string]: Record<string, string> } = {
  en,
  ja
};

const intl = createIntl({
  defaultLocale: 'en',
  locale: window.navigator.language.substring(0, 2),
  messages: translations[window.navigator.language]
});

function IntlProvider(props: Readonly<React.PropsWithChildren<unknown>>) {

  const { children } = props;

  return (
    <Provider {...intl}>
      {children}
    </Provider>
  );

}

export default IntlProvider;
