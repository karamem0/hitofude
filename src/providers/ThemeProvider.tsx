//
// Copyright (c) 2023-2024 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/hitofude/blob/main/LICENSE
//

import React from 'react';

import {
  FluentProvider as Provider,
  Theme
} from '@fluentui/react-components';

import { darkTheme, lightTheme } from '../themes/Theme';
import { InvalidOperationError } from '../types/Error';
import { ThemeName } from '../types/Model';

interface ThemeState {
  theme: Theme,
  themeName: ThemeName
}

interface ThemeContextState {
  theme: ThemeState,
  changeTheme: (themeName: ThemeName) => void
}

const ThemeContext = React.createContext<ThemeContextState | undefined>(undefined);

export const useTheme = (): ThemeContextState => {
  const value = React.useContext(ThemeContext);
  if (value == null) {
    throw new InvalidOperationError();
  }
  return value;
};

function ThemeProvider(props: Readonly<React.PropsWithChildren<unknown>>) {

  const { children } = props;

  const [ theme, setTheme ] = React.useState<ThemeState>({
    theme: lightTheme,
    themeName: ThemeName.light
  });

  const handleChangeTheme = React.useMemo(() => (value: ThemeName) => {
    switch (value) {
      case ThemeName.light:
        setTheme(() => ({
          theme: lightTheme,
          themeName: value
        }));
        break;
      case ThemeName.dark:
        setTheme(() => ({
          theme: darkTheme,
          themeName: value
        }));
        break;
      default:
        break;
    }
  }, []);

  const value = React.useMemo<ThemeContextState>(() => ({
    theme,
    changeTheme: handleChangeTheme
  }), [
    theme,
    handleChangeTheme
  ]);

  return (
    <ThemeContext.Provider value={value}>
      <Provider theme={value.theme.theme}>
        {children}
      </Provider>
    </ThemeContext.Provider>
  );

}

export default ThemeProvider;
