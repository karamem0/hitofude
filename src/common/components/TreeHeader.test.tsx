//
// Copyright (c) 2023-2025 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/hitofude/blob/main/LICENSE
//

import React from 'react';

import { render, screen } from '@testing-library/react';
import IntlProvider from '../../providers/IntlProvider';
import Presenter from './TreeHeader.presenter';
import ThemeProvider from '../../providers/ThemeProvider';
import userEvent from '@testing-library/user-event';

it('should match the snapshot when the root is true', () => {
  const params = {
    name: 'Explorer',
    menu: (
      <div data-testid="test-Menu" />
    ),
    root: true
  };
  const { asFragment } = render(
    <IntlProvider>
      <ThemeProvider>
        <Presenter {...params} />
      </ThemeProvider>
    </IntlProvider>
  );
  expect(asFragment()).toMatchSnapshot();
});

it('should match the snapshot when the root is false', () => {
  const params = {
    name: 'Explorer',
    menu: (
      <div data-testid="test-Menu" />
    ),
    root: false
  };
  const { asFragment } = render(
    <IntlProvider>
      <ThemeProvider>
        <Presenter {...params} />
      </ThemeProvider>
    </IntlProvider>
  );
  expect(asFragment()).toMatchSnapshot();
});

it('should match the snapshot when the menu is opened', async () => {
  const user = userEvent.setup();
  const container = document.body.appendChild(document.createElement('div'));
  const params = {
    name: 'Explorer',
    menu: (
      <div data-testid="test-Menu" />
    ),
    mountNode: container,
    root: true
  };
  const { asFragment } = render(
    <IntlProvider>
      <ThemeProvider>
        <Presenter {...params} />
      </ThemeProvider>
    </IntlProvider>,
    {
      container
    }
  );
  await user.click(screen.getByTitle('More option'));
  expect(asFragment()).toMatchSnapshot();
  expect(screen.getByTestId('test-Menu')).toBeInTheDocument();
});
