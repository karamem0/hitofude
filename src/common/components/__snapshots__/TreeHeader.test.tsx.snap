// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a shapshot when the root parameter is false 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: auto 1fr auto;
  grid-gap: 0.25rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: start;
  justify-content: start;
  padding: 0 0.5rem;
  margin: 1px;
  font-weight: bold;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: 1px solid transparent;
}

.emotion-0>button:last-of-type {
  display: none;
}

.emotion-0:hover>button:last-of-type {
  display: inherit;
}

.emotion-1 {
  max-width: 2rem;
  height: 2rem;
}

.emotion-2 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.emotion-3 {
  font-weight: bold;
}

<div
    class="fui-FluentProvider fui-FluentProviderr2 ___jdtuxv0_17k0bs4 f19n0e5 fxugw4r f1o700av fk6fouc fkhj508 figsok6 f1i3iumi"
    dir="ltr"
  >
    <div
      class="emotion-0"
    >
      <button
        aria-label="Back"
        class="fui-Button r1alrhcs emotion-1 ___cpe5h10_1yhfyj7 f1c21dwh f1p3nwhy f11589ue f1q5o8ev f1pdflbu fkfq4zb fjxutwb f1s2uweq fr80ssc f1ukrpxl fecsdlb f139oj5f ft1hn21 fuxngvv fwiml72 f1h0usnq fs4ktlq f16h9ulv fx2bmrt f1fg1p5m f1dfjoow f1j98vj9 f1tme0vf f4xjyn1 f18onu3q f9ddjv3 f18ktai2 fwbmr0d f44c6la"
        tabindex="0"
        title="Back"
        type="button"
      >
        <span
          class="fui-Button__icon rywnvv2"
        >
          <svg
            aria-hidden="true"
            class="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
            fill="currentColor"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10.35 3.15c.2.2.2.5 0 .7L6.21 8l4.14 4.15a.5.5 0 0 1-.7.7l-4.5-4.5a.5.5 0 0 1 0-.7l4.5-4.5c.2-.2.5-.2.7 0Z"
              fill="currentColor"
            />
          </svg>
        </span>
      </button>
      <div
        class="emotion-2"
      >
        <h3
          aria-label="Explorer"
          class="fui-Text emotion-3 ___c56hd20_1xy0w11 fk6fouc fkhj508 f1i3iumi figsok6 fpgzoln f1w7gpdv f6juhto f1gl81tg f2jf649"
          title="Explorer"
        >
          Explorer
        </h3>
      </div>
      <button
        aria-haspopup="menu"
        aria-label="More option"
        class="fui-Button r1alrhcs emotion-1 ___cpe5h10_1yhfyj7 f1c21dwh f1p3nwhy f11589ue f1q5o8ev f1pdflbu fkfq4zb fjxutwb f1s2uweq fr80ssc f1ukrpxl fecsdlb f139oj5f ft1hn21 fuxngvv fwiml72 f1h0usnq fs4ktlq f16h9ulv fx2bmrt f1fg1p5m f1dfjoow f1j98vj9 f1tme0vf f4xjyn1 f18onu3q f9ddjv3 f18ktai2 fwbmr0d f44c6la"
        id="menur3"
        role="button"
        tabindex="0"
        title="More option"
        type="button"
      >
        <span
          class="fui-Button__icon rywnvv2"
        >
          <svg
            aria-hidden="true"
            class="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
            fill="currentColor"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 5a1 1 0 1 1 0-2 1 1 0 0 1 0 2Zm0 4a1 1 0 1 1 0-2 1 1 0 0 1 0 2Zm-1 3a1 1 0 1 0 2 0 1 1 0 0 0-2 0Z"
              fill="currentColor"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should create a shapshot when the root parameter is true 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: auto 1fr auto;
  grid-gap: 0.25rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: start;
  justify-content: start;
  padding: 0 0.5rem;
  margin: 1px;
  font-weight: bold;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: 1px solid transparent;
}

.emotion-0>button:last-of-type {
  display: none;
}

.emotion-0:hover>button:last-of-type {
  display: inherit;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 2rem;
  height: 2rem;
}

.emotion-2 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.emotion-3 {
  font-weight: bold;
}

.emotion-4 {
  max-width: 2rem;
  height: 2rem;
}

<div
    class="fui-FluentProvider fui-FluentProviderr0 ___jdtuxv0_17k0bs4 f19n0e5 fxugw4r f1o700av fk6fouc fkhj508 figsok6 f1i3iumi"
    dir="ltr"
  >
    <div
      class="emotion-0"
    >
      <div
        class="emotion-1"
      >
        <svg
          fill="currentColor"
          height="1em"
          role="img"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19.453 9.95q.961.058 1.787.468.826.41 1.442 1.066.615.657.966 1.512.352.856.352 1.816 0 1.008-.387 1.893-.386.885-1.049 1.547-.662.662-1.546 1.049-.885.387-1.893.387H6q-1.242 0-2.332-.475-1.09-.475-1.904-1.29-.815-.814-1.29-1.903Q0 14.93 0 13.688q0-.985.31-1.887.311-.903.862-1.658.55-.756 1.324-1.325.774-.568 1.711-.861.434-.129.85-.187.416-.06.861-.082h.012q.515-.786 1.207-1.413.691-.627 1.5-1.066.808-.44 1.705-.668.896-.229 1.845-.229 1.278 0 2.456.417 1.177.416 2.144 1.16.967.744 1.658 1.78.692 1.038 1.008 2.28zm-7.265-4.137q-1.325 0-2.52.544-1.195.545-2.04 1.565.446.117.85.299.405.181.792.416l4.78 2.86 2.731-1.15q.27-.117.545-.204.276-.088.58-.147-.293-.937-.855-1.705-.563-.768-1.319-1.318-.755-.551-1.658-.856-.902-.304-1.886-.304zM2.414 16.395l9.914-4.184-3.832-2.297q-.586-.351-1.23-.539-.645-.188-1.325-.188-.914 0-1.722.364-.809.363-1.412.978-.604.616-.955 1.436-.352.82-.352 1.723 0 .703.234 1.423.235.721.68 1.284zm16.711 1.793q.563 0 1.078-.176.516-.176.961-.516l-7.23-4.324-10.301 4.336q.527.328 1.13.504.604.175 1.237.175zm3.012-1.852q.363-.727.363-1.523 0-.774-.293-1.407t-.791-1.072q-.498-.44-1.166-.68-.668-.24-1.406-.24-.422 0-.838.1t-.815.252q-.398.152-.785.334-.386.181-.761.345Z"
          />
        </svg>
      </div>
      <div
        class="emotion-2"
      >
        <h3
          aria-label="Explorer"
          class="fui-Text emotion-3 ___c56hd20_1xy0w11 fk6fouc fkhj508 f1i3iumi figsok6 fpgzoln f1w7gpdv f6juhto f1gl81tg f2jf649"
          title="Explorer"
        >
          Explorer
        </h3>
      </div>
      <button
        aria-haspopup="menu"
        aria-label="More option"
        class="fui-Button r1alrhcs emotion-4 ___cpe5h10_1yhfyj7 f1c21dwh f1p3nwhy f11589ue f1q5o8ev f1pdflbu fkfq4zb fjxutwb f1s2uweq fr80ssc f1ukrpxl fecsdlb f139oj5f ft1hn21 fuxngvv fwiml72 f1h0usnq fs4ktlq f16h9ulv fx2bmrt f1fg1p5m f1dfjoow f1j98vj9 f1tme0vf f4xjyn1 f18onu3q f9ddjv3 f18ktai2 fwbmr0d f44c6la"
        id="menur1"
        role="button"
        tabindex="0"
        title="More option"
        type="button"
      >
        <span
          class="fui-Button__icon rywnvv2"
        >
          <svg
            aria-hidden="true"
            class="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
            fill="currentColor"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 5a1 1 0 1 1 0-2 1 1 0 0 1 0 2Zm0 4a1 1 0 1 1 0-2 1 1 0 0 1 0 2Zm-1 3a1 1 0 1 0 2 0 1 1 0 0 0-2 0Z"
              fill="currentColor"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
